### ISSUE TITLE
Support workspace crate version denylist / local yanking

### ISSUE THREAD

**Issue by demurgos** [rust-lang/cargo#13856]
### Problem

Occasionally, crates release a semver-violating version that causes breakage when consumed.

Cargo currently has two features to address this issue:
1. Crate authors can mark the offending crate as "yanked"
2. Crate consumers can use a lock file with a pinned working version

The first approach is the best one, but it requires active intervention of the crate authors which may sometime take a few days. The second approach works great if the dependency was already used through a good version, however it's not easy to use when we want start using this dependency. This dependency can be deep in the chain so applying requirements locally in the workspace may not work. There also solutions involving forking the crate or pointing to some other source (such as the Git repo), but these feel more like a workarounds since the working version already exists in the registry, cargo resolution should be able to resolve it.

The scenario is the following:
1. `foo` has a private dependency on `clumsy ^1.0.0`
2. `bar` has a private dependency on `clumsy ^2.0.0`
4. `myproject` depends on `foo` only
5. `clumsy` releases `2.0.1` which contains a semver violation
6. `myproject` wants to add a dependency on `bar`, but by doing so cargo pulls the broken `clumsy 2.0.1` instead of the working `clumsy 2.0.0`

`myproject` now needs to needs to edit the lock file to add `foo` and `clumsy 2.0.0`, which is pretty complex.

### Proposed Solution

The solution for me would be to improve help resolution / writing the lock file in presence of semver violations without expecting coordination from third-party library, In the problem description, the fix could also be applied in `bar`, but in all cases `myproject` has to wait for a fix upstream instead of easily fixing it itself locally. I expect the solution to be applied at the workspace level and work with `Cargo.lock` or `patch` sections.

1. The restricted solution for this issue of semver violations would be to have a way to mark at the workspace level that a crate version should be treated as yanked. When cargo performs resolution, a crate version is treated as yanked if it is marked as such either on the registry or in the workspace manifest (for example in the patch section). Resolution proceeds with regular yanking semantics. With this solution, `myproject` would flag `clumsy@2.0.1` as yanked and cargo resolution would select `clumsy@2.0.0`.

2. A more general solution would be to allow patching dependency requirements. `myproject` could patch `bar` to depend on `clumsy =2.0.0` instead of `clumsy ^2.0.0`.

3. Another general solution would be support for version substitution when patching the registry. Right now `patch.crates-io.clumsy` can swap it to a `path` or `git` dependency, but it could also support swapping for a _different_ version from the registry.

### Notes

Despite being currently affected by scenario above, I prefer to avoid sharing exact details of the crates involved as general improvements to occasional semver violations feel more important than blaming specific instances which should be fixed eventually anyway. It's not the first time that I encounter a semver violation, nor the last and it would be nice to have an escape hatch.

**Comment by mladedav** [comment]
You currently also have other options:
- Downgrade the package without manually editing the lock file. You can run `cargo update -p clumsy@2.0.1 --precise 2.0.0 to downgrade
- Pin the version in your `Cargo.toml` with something like `bar = "=2.0.0"`. This has the drawback of missing potential upgrades after the broken version, but then again, if the crate broke semver you might want to check the future `2.0.2` manually first and not allow automatic updates. You would have to add the dependency directly in one of your projects though to pin the version.

Are those solutions good enough for you?

**Comment by weihanglo** [comment]
Thanks for the report. I love the amount of details you filled in!

I wonder if we could consolidate similar feature requests into one location, so people don't need to jump between issues to get the full picture. Personally I favor #5286 and close this. Does that sound good to you?

**Comment by demurgos** [comment]
@weihanglo 
I don't think that these issues cover the same use cases. As far as I understand, #5286 is about version requirements (a local property of a crate) while this issue is about resolution (a global property of the dependency tree).

@mladedav
These are good ideas using the current Cargo features (I did not try them though), but I still think that local yanking would be a useful feature.

**Comment by mladedav** [comment]
4th comment in #5286:

> Scenario: Package A depends on B which depends on C. C pushes a broken version. B has not yet pushed a blacklist for the broken C. A should be able to blacklist it without a dependency on C.

I appreciate it's not the issue itself, but this seems to exactly describe what you're proposing here?

**Comment by demurgos** [comment]
I see, yes this comment matches the goal of the issue here; but then @epage replies that the global denylist is not the goal of #5286 so I find it a bit confusing. When searching for existing issues I saw this one, and their comment is actually why I opened a dedicated issue for this feature. Anyway, if you think that it's clear enough if this issue is closed/it's better for triage then it's fine for me. :+1:

**Comment by epage** [comment]
While I focused on the implementation in my discussion for #5286, I think the most important thing is to focus on the underlying requirements and to allow discussion of multiple solutions for the underlying requirements in one place to avoid fracturing the discussion.

However, I think the solutions discussed in each are tightly related to the underlying problem in each which are distinct.

In #5296 the concern is with the published result of a package (in this issue's terms, from the perspective of the publisher of `bar`).  The only thing that affects dependents on a published package are version requirements.

This issue is about the generation of `my-project`s `Cargo.lock` without the cooperation of `bar` or `clumsy`.

The user and who they are trying to help in each story is different.

**Comment by epage** [comment]
This issue is about controlling what versions are allowed.  #15824 is about controlling what packages are allowed.  They likely should be designed together.
